fix behavior of scipy.linalg.block_diag for empty input

levelfour opened the pull request:
I fix the bug #4596. This bug happens because ``np.array([[]]).shape`` is actually ``(1,0)``. The code expected ``(0,0)``.

jaimefrio commented on scipy/linalg/special_matrices.py:
```diff
@@ -528,7 +528,7 @@ def block_diag(*arrs):
                          "greater than 2: %s" % bad_args)
 
-    shapes = np.array([a.shape for a in arrs])
+    shapes = np.array([a.shape if np.array(a.shape).all() else [0, 0] for a in arrs])
```
Rather than iterating over the shape with `all`, it is probably going to be clearer if you do the check on `a.size`. I would also explicitly check for inequality to `0`, but that may just be me, i.e.:

```
shapes = np.array([a.shape if a.size > 0 else (0, 0) for a in arrs])
```

levelfour replied:
@jaimefrio Thanks for your comment. And you are right. I fixed and pushed it.

jaimefrio commented:
You will eventually be asked to squash your commits into a single one, see [here](http://docs.scipy.org/doc/numpy/reference/generated/numpy.searchsorted.html), you might as well go ahead and do it now. When you do so, modify your commit message so that it includes the text "`fixes #4596`", so that the corresponding issue gets closed automatically by github.

Other than that, LGTM.

ev-br commented:
Can you add a regression test please.

levelfour commented:
@jaimefrio According to your advice, I squashed my commits.

@ev-br If I would like to add a test, is it right to add into scipy/linalg/test/test_special_matrices.py?

ev-br commented:
Yes, there is even `TestBlockDiag` class in `linalg/tests/test_special_matrices`. I'd add a new method to this test class.
Please also add a comment with a reference to the github issue and a short explanation. Something along the lines of `# regression test for gh-4569: check the shape of the result for empty matrix inputs` or something in that ballpark.

(Incidentally, this is an old-style class on python2, so one may want to fix that as well --- this is best done in a separate PR and would require checking that the number of tests run is the same)

Also, please make sure that your commit messages are in the format described in 
http://docs.scipy.org/doc/numpy/dev/gitwash/development_workflow.html --- in this case I'd use, for instance, "BUG" for the prefix.

levelfour commented:
Thanks for your attentive comments. I have no experience for contribution, so I'm not so sure if my way is correct.
According to your comments, I added regression test and fixed commit message.
I'd appreciate if you could take time to check them.

ev-br commented:
Merged in https://github.com/scipy/scipy/commit/2b856a13d9451a4d1c7d85557b9be2068c0adf0f wih minor style tweaks. Thanks @levelfour!

argriffing commented:
This PR appears to treat all size-zero matrices equally
```python
shapes = np.array([a.shape if a.size > 0 else [0, 0] for a in arrs])
```
but issue https://github.com/scipy/scipy/issues/4908 wants to distinguish between matrices of shape (0, 0) vs. matrices of shape (0, 2), for example.

ev-br commented:
@argriffing Do you mean that you see a use for distinguishing between different zero-size matrices?

argriffing commented:
> Do you mean that you see a use for distinguishing between different zero-size matrices?

I don't think this is controversial, if you mean something like distinguishing among zero-size matrices with shapes like (0, 5) or (10, 0) or (0, 0).  I remember seeing a web page related to this, maybe for Matlab or Mathematica, explaining how awesome this is but I cannot find or remember the link.
